dotnet/Razorvine.Pyrolite/Pyrolite/Pickle/ReadOnlyMemoryReader.cs

```diff
@@ -0,0 +1,56 @@
+﻿using System;
+
+namespace Razorvine.Pickle
+{
+    internal struct ReadOnlyMemoryReader : IInputReader
+    {
+        private readonly ReadOnlyMemory<byte> input;
+        private int position;
+
+        public ReadOnlyMemoryReader(ReadOnlyMemory<byte> bytes)
+        {
+            input = bytes;
+            position = 0;
+        }
+
+        public byte ReadByte()
+        {
+            return input.Span[position++];
+        }
+
+        public ReadOnlySpan<byte> ReadBytes(int bytesCount)
+        {
+            var result = input.Slice(position, bytesCount).Span;
+            position += bytesCount;
+            return result;
+        }
+
+        public string ReadLine(bool includeLF = false)
+        {
+            return PickleUtils.rawStringFromBytes(ReadLineBytes(includeLF));
+        }
+
+        public ReadOnlySpan<byte> ReadLineBytes(bool includeLF = false)
+        {
+            var result = ReadBytes(GetLineEndIndex(includeLF));
+            if (!includeLF)
+                Skip(1);
+            return result;
+        }
+
+        public void Skip(int bytesCount)
+        {
+            position += bytesCount;
+        }
+
+        private int GetLineEndIndex(bool includeLF = false)
+        {
+            var bytes = input.Slice(position).Span;
```

Comment:
You may want to get the `.Span` first, and then `.Slice`.

See Tip 1 under Tricks of the Trade in [https://www.codemag.com/Article/1807051/Introducing-.NET-Core-2.1-Flagship-Types-Span-T-and-Memory-T](https://www.codemag.com/Article/1807051/Introducing-.NET-Core-2.1-Flagship-Types-Span-T-and-Memory-T)

Reply:
Great catch! I did not know about it.

Reply:
Yep. I have a prototype code analyzer that catches the two cases (tips shown in the article), if you want to try it: https://github.com/dotnet/corefxlab/tree/master/samples/SpanUsage